Objects/genobject.c

```diff
@@ -158,6 +159,15 @@ gen_dealloc(PyGenObject *gen)
     PyObject_GC_Del(gen);
 }
 
+static void
+gen_dealloc(PyGenObject *gen)
```

Comment:
This doesn't seem right to me. `tp_dealloc` can't be called concurrently on an object.

Reply:
In this case the racing part is mainly `_PyFrame_ClearExceptCode` and setting the generator's flags. `_PyEval_FrameClearAndPop` can race with this as it also clears a generator, but instead it's done by `ceval.c`.

Reply:
I don't understand -- how can `_PyFrame_ClearExceptCode` race with `gen_dealloc`?

It looks like `gen_dealloc` is only called from the `tp_dealloc` handler. If an object is being destroyed, then no other thread has access to it (or there's a separate serious bug).

Reply:
Hmm I'm not too sure either. Let me revert this fix for now and try to hunt it down again.